### PR TITLE
Unstructured OpenMP code

### PR DESCRIPTION
**Pull request by vdonaldson**
Address several problems with OpenMP constructs and unstructured code.

In fir, code that contains a GOTO, EXIT, or any of a number of other
branches is _unstructured_.  OpenMP generally prohibits branching into
or out of an OpenMP construct, but allows unstructured branches where
the source and target are both local to the construct.  A structured
loop is implemented with a fir.do_loop op, and a structured IF is
implemented with a fir.if op.  Unstructured loops and IFs are implemented
with explicit branches between basic blocks.  This PR allows an OpenMP
construct to immediately follow unstructured code (see PR 1077), and
allows an OpenMP construct to contain unstructured code (see Issue 1120).

The same issues are likely present in OpenACC code.  The infrastructure
changes in this PR should also be valid for OpenACC code, but file
OpenACC.cpp is not changed.

### CODE REVIEW

**Review by jeanPerier — APPROVED** [review]
It looks reasonable to me. Please wait for a more enlightened opinion regarding these CFGs/OpenMP changes like @schweitzpgi and @kiranchandramohan.

**Comment by Leporacanthicus** [comment]
I'm trying this patch now.

I hit this:
```
bbc: /data/f18-llvm-project/flang/lib/Lower/OpenMP.cpp:117: void createEmptyRegionBlocks(fir::FirOpBuilder &, std::list<Fortran::lower::pft::Evaluation> &): Assertion `eval.block->empty() && "block is not empty"' failed.
```
I'm attempting to produce a small reproducer. I'm not suggesting this is a blocker for getting this patch in, but that there may be some more work needed for this to be completely solved - up to others to decide whether we should "work more on this patch" or "add separate patch later".

**Comment by Leporacanthicus** [comment]
Here's a small reproducer:
```
program lineloop
  implicit none
  INTEGER ::  n
  
  !$OMP PARALLEL DO NUM_THREADS(2) SCHEDULE(STATIC,1) PRIVATE(n)
  DO n = 1, 2
     IF ( n > 2 )  CYCLE
  END DO
  !$OMP END PARALLEL DO
end program lineloop
```

Note that a previously posted program with `print *,i` after the `if  ... cycle` works fine - and of course, if I add some print to this code, it also compiles without issue. The original code is in mkba_sweep.f90 in the SNAP application, and the original loop (`line_loop`) is just under 300 lines of code.

**Comment by Leporacanthicus** [comment]
Good news: As far as I can tell, the remainder of the SNAP code compiles and builds correctly with this patch.

File: flang/test/Lower/OpenMP/omp-unstructured.f90 [hunk 1]
```diff
@@ -0,0 +1,111 @@
+! RUN: bbc %s -fopenmp -o "-" | FileCheck %s
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/test/Lower/OpenMP/omp-unstructured.f90:3]: I've added comments for each test subroutine.  Subroutine ss1 exercises the 1077 functionality.

File: flang/test/Lower/OpenMP/omp-unstructured.f90 [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/test/Lower/OpenMP/omp-unstructured.f90:4]: I've added a comment

File: flang/test/Lower/OpenMP/omp-unstructured.f90 [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  do i = 1, 3
+    if (i .eq. n) exit
+    print*, 'ss1-A', i
+  enddo
+  !$omp master
+    print*, 'ss1-B', i
+  !$omp end master
+  print*
+end
+
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/test/Lower/OpenMP/omp-unstructured.f90:31]: I've added a comment

File: flang/test/Lower/OpenMP/omp-unstructured.f90 [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    print*, 'ss2-A', n
+    do i = 1, 3
+      if (i .eq. n) exit
+      print*, 'ss2-B', i
+    enddo
+  !$omp end master
+  print*, 'ss2-C', i
+  print*
+end
+
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/test/Lower/OpenMP/omp-unstructured.f90:60]: I've added a comment

File: flang/lib/Lower/OpenMP.cpp [hunk 5]
```diff
@@ -93,9 +93,27 @@ static void genObjectList(const Fortran::parser::OmpObjectList &objectList,
   }
 }
 
+/// Create empty blocks for the current region.
+/// These blocks replace blocks parented to an enclosing region.
+void createEmptyRegionBlocks(
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/OpenMP.cpp:98]: It is intended to operate on the current region, as updated before the call in createBodyOfOp.

**Comment by vdonaldson** [comment]
> I think I can close #1077 in favour of this patch. Is that OK?

That was the intent of adding a more narrowly focused implementation of the 1077 code here, along with some related comment updates.  We've still got problems to solve though.

**Comment by vdonaldson** [comment]
> Here's a small reproducer: . . .

Thanks for this small test case.  It turns up at least one more problem variation, and hits at the intersection of some of the other known issues.  The most obvious problem is that the `$omp parallel do` effectively splits into `$omp parallel` + `$omp do`, each with its own region.  The code as is doesn't allow for that.  There is also an instance of the problem of requiring a construct exit target inside the correct region that isn't accounted for, and possibly other issues as well.  I don't yet have a fix for this set of problems.

The test case as is hits a `CYCLE to construct outside of PARALLEL DO construct is not allowed` error.  That is in invalid error that is easy to work around; you may already have a fix for that problem.

File: flang/lib/Lower/PFTBuilder.cpp [hunk 6]
```diff
@@ -310,6 +310,17 @@ class PFTBuilder {
 
   void exitConstructOrDirective() {
     rewriteIfGotos();
+    auto *eval = constructAndDirectiveStack.back();
+    if (eval->isExecutableDirective()) {
+      // A construct at the end of an (unstructured) OpenACC or OpenMP
+      // construct region must have an exit target inside the region.
```
[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/PFTBuilder.cpp:316]: This code is exercised by test subroutines ss2 and ss3.

**Comment by Leporacanthicus** [comment]
> > Here's a small reproducer: . . .
> 
> Thanks for this small test case. It turns up at least one more problem variation, and hits at the intersection of some of the other known issues. The most obvious problem is that the `$omp parallel do` effectively splits into `$omp parallel` + `$omp do`, each with its own region. The code as is doesn't allow for that. There is also an instance of the problem of requiring a construct exit target inside the correct region that isn't accounted for, and possibly other issues as well. I don't yet have a fix for this set of problems.

Feel free to work on that in a separate patch. I can see how this patch helps a whole lot, and I don't know enough about this subject to say which is the better approach - two patches, or try to fix everything in one patch.
> 
> The test case as is hits a `CYCLE to construct outside of PARALLEL DO construct is not allowed` error. That is in invalid error that is easy to work around; you may already have a fix for that problem.

Yes, I have reverted the commit that causes that check - it hits a couple of different places in SNAP. I think there is an updated patch in review at the moment.

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Thanks, Val for this patch. I guess we can fix the other issue in a separate patch.

**Review by Leporacanthicus — APPROVED** [review]
LGTM.

I have raised #1192 for the remaining  known issue in SNAP.

With this fix, I'm able to compile all of SNAP, and with four $OMP directives commented out, the application runs to completion with what appears to be correct results.

**Comment by vdonaldson** [comment]
I've pushed a change that takes a step toward resolving the issue with (implicit or explicit) nested constructs.  A full solution to that problem may need special code for some cases, such as the implicit `parallel do` -> `parallel` + `do` nesting case.  The failure symptom for the test case above changes to a missing basic block terminator.

As a general note, unstructured code (in the fir sense) in OpenMP code is probably not all that common.  One way to artificially "create" more tests that exercise unstructured OpenMP code generation is to use the `-no-structured-fir` developer option.  That option bypasses the opportunity to generate _fir.do_loop_ and _fir.if_  ops when that is otherwise allowed, in favor of unstructured branches between basic blocks.

I'll merge this PR as a base for OpenMP developers to address the remaining issues.

**Pull request merged**